### PR TITLE
Use Conduit-typed messages

### PR DESCRIPTION
**Pull request by christophermaier**
Depends on https://github.com/operable/conduit/pull/1; when that's merged, `mix.exs` in this PR will need to be updated accordingly.

### CODE REVIEW

File: lib/carrier/messaging/connection.ex [hunk 1]
```diff
@@ -94,10 +94,14 @@ defmodule Carrier.Messaging.Connection do
...
-  def publish(conn, message, kw_args) when is_map(message) do
+
+  # Here, we assume we're being passed a Conduit-enabled struct
+  # (aside: any way to verify that statically?) We'll do the encoding
+  # to JSON internally
+  def publish(conn, %{__struct__: _}=message, kw_args) do
     topic = Keyword.fetch!(kw_args, :routed_by)
 
-    encoded = Poison.encode!(message)
+    encoded = message.__struct__.encode!(message)
```
Comment (kevsmith) [inline lib/carrier/messaging/connection.ex]: Might be nice to pull this into Conduit, no? Then we could check to see if the struct is Conduit-based or not and return an appropriate error.

```elixir
Conduit.encode!(message)
```

File: lib/cog/adapter.ex [hunk 2]
```diff
@@ -31,35 +31,31 @@ defmodule Cog.Adapter do
...
 
       def handle_call({:receive_message, sender, room, message, id, initial_context}, _from, state) do
-        message = payload(sender, room, message, id, initial_context)
+        message = %Cog.Messages.AdapterRequest{id: id,
+                                               sender: sender,
+                                               room: room,
+                                               text: message,
+                                               initial_context: initial_context,
+                                               adapter: name(),
+                                               module: to_string(__MODULE__),
```
Comment (kevsmith) [inline lib/cog/adapter.ex]: I really really really want this to die die die :)

**Comment by kevsmith** [comment]
I think some/a lot of this new code will change when we integrate adapters v2 but this is a great start.

👍  to merge

File: lib/carrier/messaging/connection.ex (hunk 1 above)
Comment (christophermaier) [inline lib/carrier/messaging/connection.ex]: Agreed; added #902 to capture this.

File: lib/cog/adapter.ex (hunk 2 above)
Comment (christophermaier) [inline lib/cog/adapter.ex]: indeed

**Pull request merged**